File: app/schemas/access_event.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.vehicle_access_event import AccessDecision, AccessDirection, ProcessingStatus, RecognitionSource
from app.schemas.pagination import PaginationMeta


class AccessEventCreateBase(BaseModel):
    parking_lot_id: int
    direction: AccessDirection
    plate_number: str = Field(min_length=1, max_length=64)


class AccessEventRecognizeIn(BaseModel):
    parking_lot_id: int
    direction: AccessDirection
    image_token: str | None = None
    plate_number_hint: str | None = None


class AccessEventManualIn(AccessEventCreateBase):
    recognition_confidence: float | None = Field(default=None, ge=0, le=1)


class AccessEventOut(BaseModel):
    id: int
    parking_lot_id: int
    parking_spot_id: int | None
    booking_id: int | None
    user_id: int | None
    vehicle_id: int | None
    plate_number: str
    normalized_plate_number: str
    direction: AccessDirection
    recognition_confidence: float | None
    recognition_source: RecognitionSource
    image_url: str | None
    video_url: str | None
    frame_timestamp: float | None
    processing_status: ProcessingStatus
    decision: AccessDecision
    reason: str
    created_at: datetime

    class Config:
        from_attributes = True


class RecognitionCandidateOut(BaseModel):
    plate: str
    normalized_plate: str
    confidence: float
    valid: bool
    reason: str


class AccessEventRecognitionOut(AccessEventOut):
    raw_text: str | None = None
    candidates: list[RecognitionCandidateOut] = Field(default_factory=list)
    provider: str | None = None
    confidence: float | None = None
    recognition_reason: str | None = None
    processing_status_detail: str | None = None
    selected_plate: str | None = None
    normalized_plate: str | None = None
    preprocessing_steps: list[str] = Field(default_factory=list)


class AccessEventListResponse(BaseModel):
    items: list[AccessEventOut]
    meta: PaginationMeta

File: frontend/src/shared/types/access-event.ts
import type { ListResponse } from './common';

export type AccessDirection = 'entry' | 'exit';
export type RecognitionSource = 'manual' | 'mock' | 'provider';
export type AccessDecision = 'allowed' | 'denied' | 'review';

export interface AccessEvent {
  id: number;
  parking_lot_id: number;
  parking_spot_id: number | null;
  booking_id: number | null;
  user_id: number | null;
  vehicle_id: number | null;
  plate_number: string;
  normalized_plate_number: string;
  direction: AccessDirection;
  recognition_confidence: number | null;
  recognition_source: RecognitionSource;
  recognition_provider: string | null;
  recognition_diagnostics: Record<string, unknown> | null;
  image_url: string | null;
  video_url: string | null;
  frame_timestamp: number | null;
  processing_status: "pending" | "processed" | "failed";
  decision: AccessDecision;
  reason: string;
  created_at: string;
}

export interface AccessEventManualPayload {
  parking_lot_id: number;
  direction: AccessDirection;
  plate_number: string;
  recognition_confidence?: number;
}

export interface AccessEventRecognizePayload {
  parking_lot_id: number;
  direction: AccessDirection;
  image_token?: string;
  plate_number_hint?: string;
}

export interface AccessEventsQuery {
  parking_lot_id?: number;
  plate_number?: string;
  decision?: AccessDecision;
  direction?: AccessDirection;
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
}

export type AccessEventListResponse = ListResponse<AccessEvent>;
